src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -1339,6 +1348,13 @@ private static unsafe string UInt64ToDecStr(ulong value, int digits)
                 digits = 1;
 
             int bufferLength = Math.Max(digits, FormattingHelpers.CountDigits(value));
+
+            // For single-digit values that are very common, especially 0 and 1, just return cached strings.
+            if (bufferLength == 1)
```

Comment:
Skip range check for this one (moving `bufferLength` after loop) and use 
```csharp
string[] ditgitCache = s_singleDigitStringCache;
if ((uint)value < (uint)ditgitCache.Length)
{
    return ditgitCache[value];
}
```
Would also allow you to have any number in cache? (within reason)

Reply:
> Would also allow you to have any number in cache? (within reason)

As noted earlier, it would also need to check `digits`.  That's why I did the comparison against `bufferLength`.

Reply:
Ahh... leading zeros...